src/binary_annotator.cpp

```diff
@@ -0,0 +1,647 @@
+            BuildUnion(union_offset, realized_type, regions, field);
+
+        regions.emplace_back(MakeBinaryRegion(
+            offset, sizeof(uint32_t), BinaryRegionType::UOffset, 0,
+            union_offset,
+            std::string("offset to union[") + std::to_string(i) + "] (`" +
+                enum_type + "`)"));
+
+        offset += sizeof(uint32_t);
+      }
+    } break;
+    default: {
+      if (IsScalar(field->type()->element())) {
+        const BinaryRegionType binary_region_type =
+            GetRegionType(field->type()->element());
+
+        const uint64_t type_size = GetTypeSize(field->type()->element());
+
+        // TODO(dbaileychess): It might be nicer to user the
+        // BinaryRegion.array_length field to indicate this.
+        for (size_t i = 0; i < vector_length; ++i) {
+          regions.emplace_back(MakeBinaryRegion(
+              offset, type_size, binary_region_type, 0, 0,
+              std::string("value[") + std::to_string(i) + "]"));
+          offset += type_size;
+        }
+      }
+    } break;
+  }
+
+  sections_.insert(std::make_pair(
+      vector_offset,
+      MakeBinarySection(
+          std::string(table->name()->c_str()) + "." + field->name()->c_str(),
+          BinarySectionType::Vector, std::move(regions))));
+}
+
+std::string BinaryAnnotator::BuildUnion(uint64_t offset,
+                                        const uint8_t realized_type,
+                                        std::vector<BinaryRegion> &regions,
+                                        const reflection::Field *field) {
+  const reflection::Enum *next_enum =
+      schema_->enums()->Get(field->type()->index());
+
+  const reflection::EnumVal *enum_val = next_enum->values()->Get(realized_type);
+
+  const reflection::Type *union_type = enum_val->union_type();
+
+  if (union_type->base_type() == reflection::BaseType::Obj) {
+    const reflection::Object *object =
+        schema_->objects()->Get(union_type->index());
+
+    if (object->is_struct()) {
+      offset = BuildStruct(offset, regions, object);
+    } else {
+      BuildTable(offset, BinarySectionType::Table, object);
+    }
+  }
+  // TODO(dbaileychess): handle the other union types.
+
+  return enum_val->name()->c_str();
+}
+
+void BinaryAnnotator::InsertFinalPadding() {
+  uint64_t offset = 0;
+  BinarySection *previous_section = nullptr;
+
+  for (auto &current_section : sections_) {
+    BinarySection &section = current_section.second;
+    const uint64_t section_start_offset = current_section.first;
+    const uint64_t section_end_offset =
+        section.regions.back().offset + section.regions.back().length;
+
+    if (offset < section_start_offset && previous_section != nullptr) {
+      // We are at an offset that is less then the current section. So padding
+      // has to be added to the previous section (if it exists).
+      const uint64_t pad_bytes = section_start_offset - offset + 1;
+
+      previous_section->regions.emplace_back(
```

Comment:
If the binary is from a newer version of the schema being used here, there may be data in between sections. That's not strictly padding, so consider leaving a TODO to handle this case.

We can assume it's padding if it's 
1. only zeroes
2. the next section is aligned appropriately
3. there are fewer padding bytes than the alignment of the next section (e.g. if the next section is aligned to 8 there should be at most 7 padding bytes).

Reply:
Good point. The vtable itself has the size of the vtable, so I known from just reading the binary how many fields are present.  So I will update the code to say these fields are "offsets to an `unknown data`"